### PR TITLE
test(#5987): add behaviour coverage for code applier branch handling

### PR DESCRIPTION
**Pull request by waynesun09**
## Summary

Adds `e2e/behaviour` scenarios that assert the code applier's branch handling end to end against live GitHub, plus the primitives the suite was missing to express them (issue #5987). Until now every scenario stopped at dispatch/labeling; nothing drove a run through the post-scripts to a real push and asserted the resulting branch state.

The scenarios are mechanically gated: they carry `@requires:capability:applier-branch-namespace` and are skipped unless the runner declares that capability via `BEHAVIOUR_CAPABILITIES`, because they assert applier behavior that ships with the upcoming agents-side branch-namespace enforcement. CI stays green until the pinned agents release includes it; then the capability gets enabled.

## Related Issue

Refs #5987 (delivers the primitives and scenarios A/C; see "Deviation" below for scenario B)

## Changes

- **`checkout_branch` dummy-runtime op** (`internal/runtime/dummy.go`): takes a single regex-validated branch name. Probes the remote with `git ls-remote --exit-code --heads` — a missing ref (`rc=2`) falls back to branching off `HEAD`, any other probe failure fails the op instead of silently degrading. When the ref exists it's fetched via the same `refs/heads/` scope (so a same-named tag can never be resolved instead — regression-tested) and the branch is based on `FETCH_HEAD`. The op then records one marker commit, so the applier post-script has content to push **and** a wrongful push visibly moves the target branch tip. Covered by mock-exec unit tests plus real-shell/git integration tests for the existing-ref, missing-ref, broken-remote, and tag-collision paths.
- **Branch assertion steps** (`pkg/behaviourtest/steps/branch.go`):
  - `an open pull request on branch "<name>"` / `a remote branch "<name>" seeded with a commit` — deterministic Given fixtures that tolerate pool-repo debris; PR base resolved via `GetDefaultBranch`.
  - `the tip of branch "<name>" is recorded` / `branch "<name>" is unchanged` — capture `GetBranchRef` before the run, re-check after.
  - `the pull request head branch matches "<pattern>"` — exactly one open PR head must match the anchored Go-regex pattern; supports an `<issue>` placeholder (also honored in `checkout_branch` table args — scoped to that op only).
  - `the harness "<agent>" workflow fails reporting "<text>"` — waits for a terminally-failed harness run, then asserts the post-script failure comment on the scenario PR; context-cancellable end to end.
- **Capability gating**: new `@requires:capability:<name>` compatibility tag recognized by `SkipErrorForTagNames`, wired to the comma-separated `BEHAVIOUR_CAPABILITIES` env var. A malformed tag (empty capability name) is a hard error, not a silently-permanent skip.
- **Driver surface**: `scm.Driver` gains `ListOpenChangeProposals`/`ListComments`; `ci.Driver` gains `WaitForFailedHarnessAgent` using artifact-first detection (`fullsend-<agent>` uploads with `if: always()`, resolving standard stage jobs like `Fix`/`Code` too, not just custom-harness matrix jobs), with a job-name fallback that scans every completed run — not just ones already known to have failed — so a successful job still fails fast through that path too. Surfaced job-listing errors on timeout, unit tests for both detection paths. All three methods documented in the behaviour-testing "API changes" section.
- **Feature file** `e2e/behaviour/features/code/branch-namespace.feature`: scenario A (foreign branch renamed into the issue namespace, decoy tip unchanged; decoy uses issue number 990000099 so it can never collide with a real issue's anchored pattern) and scenario C (conforming `agent/<issue>-<slug>` branch pushed with no rename).
- **Cleanup**: scenario-created branches/PRs are tracked and removed in `CleanupScenario`, plus a namespace sweep — gated on the scenario's issue number alone, not on whether a branch step happened to run — that reclaims applier-created `agent/<issue>-*` PRs even when the scenario fails before its head-match assertion. Deduplicated against PRs the assertion step already tracked, so a passing run doesn't double-close/delete.

## Deviation from the issue plan

Issue #5987's scenario B (fix run refuses to push from a non-matching branch) is **not expressible in this suite today**: the fix stage's only dispatch route is a `changes_requested` review submitted by the org review bot (`reusable-dispatch.yml`), and suite-posted comments are bot-authored, which both the per-repo shim and the `/fs-fix` route drop. The failure-assertion step and `WaitForFailedHarnessAgent` are landed and unit-tested as ready primitives, the limitation is documented in the feature file and suite docs, and the fix-side refusal stays covered by script-level tests in the agents repo until a suite-reachable fix trigger exists.

## Testing

- [x] `make lint` passes (staged first)
- [x] `go test ./internal/runtime/ ./pkg/behaviourtest/...` — unit tests for the op (valid/invalid names, exec failure, real-shell integration for both basing paths, the fail-closed remote-error path, and the tag-collision regression), every new step, the capability gate (including the malformed-tag error path), the cleanup sweep (including dedup and no-branch-step gating), and both `WaitForFailedHarnessAgent` detection paths
- [ ] Live `make behaviour-test` with `BEHAVIOUR_CAPABILITIES=applier-branch-namespace` — pending the agents-side branch-namespace enforcement in the pinned agents release; without the capability the scenarios skip (verified by the gating unit tests)

The two pre-existing `TestDummyRuntime_Bootstrap`/`TestDummyRuntime_ClearIterationArtifacts` failures on machines with a live local sandbox are unrelated (also fail on clean `main`).

This PR went through two rounds of a 3-agent review squad (Claude, Grok). Round 1 found the fix-stage scenario was undispatchable, the failure waiter only matched custom-harness job names, tip-unchanged assertions were vacuous, and the branch-checkout fetch failed open on transient errors — all fixed above. Round 2 caught a cleanup double-close/delete regression, a fragile sweep gate, a fetch/tag-collision edge case, and a fast-fail regression in the job-name fallback — also fixed and covered by new regression tests.

## Checklist

- [x] PR title follows [Conventional Commits](COMMITS.md)
- [x] Commits are signed off (DCO)
- [x] I wrote this contribution myself and can explain all changes in it

### CODE REVIEW

**Review by rh-hemartin — APPROVED** [review]
Looks good to me.

File: internal/runtime/dummy.go [hunk 1]
```diff
@@ -254,6 +261,31 @@ func executeBehaviourOp(rt DummyRuntime, sandboxName, repoDir string, op Behavio
...
+		}
+		// Fetch the remote ref first so the checked-out branch carries the
+		// remote branch's commits when it exists (FETCH_HEAD basing). When
+		// the ref does not exist remotely, fall back to branching off the
+		// current HEAD. Deliberately a single narrow capability — not a
+		// general shell op.
+		quoted := shellQuote(name)
+		cmd := fmt.Sprintf(
+			"cd %s && if git fetch origin %s 2>/dev/null; then git checkout -B %s FETCH_HEAD; else git checkout -B %s; fi",
+			shellQuote(repoDir), quoted, quoted, quoted)
```
[the remark this answers is not in the document]
Comment (waynesun09) [reply internal/runtime/dummy.go]: Already addressed as of ee5a2baf — this review ran against the initial push (2036c475), before either fix round. `checkout_branch` now probes with `git ls-remote --exit-code --heads origin <branch>`: exit 2 (ref not found) falls back to HEAD, any other non-zero exit (network/auth) fails the op with the exit code preserved in the error message instead of silently falling back. Covered by `TestExecuteBehaviourOp_CheckoutBranchRealShellRemoteError` (broken remote → op errors) and the missing/existing-ref real-shell tests in internal/runtime/dummy_test.go.

File: docs/guides/dev/behaviour-drivers.md [hunk 2]
```diff
@@ -47,7 +47,7 @@ Use `forge.Client` for operations it already exposes; add REST helpers inside th
 
 ## Adding a CI driver
 
-1. Implement `ci.Driver` — `WaitForWorkflow`, `FindCompletedWorkflowRun`, `AssertNoWorkflow`, `GetRunLogs`, `DownloadArtifacts`, `DownloadNamedArtifactFromRun`, `DownloadNamedArtifactAfter`, `WaitForHarnessAgent`, `AssertNoHarnessAgentArtifact`, `CountHarnessDispatches`.
+1. Implement `ci.Driver` — `WaitForWorkflow`, `FindCompletedWorkflowRun`, `AssertNoWorkflow`, `GetRunLogs`, `DownloadArtifacts`, `DownloadNamedArtifactFromRun`, `DownloadNamedArtifactAfter`, `WaitForHarnessAgent`, `WaitForFailedHarnessAgent`, `AssertNoHarnessAgentArtifact`, `CountHarnessDispatches`.
```
[the remark this answers is not in the document]
Comment (waynesun09) [reply docs/guides/dev/behaviour-drivers.md:50]: This flags docs/guides/dev/ itself, not something introduced by this PR — that directory (behaviour-testing.md, behaviour-drivers.md, cli-internals.md, e2e-testing.md, testing-workflows.md, tracing.md) predates this PR on main and is the established location for contributor/dev-facing docs, distinct from the admin/user guide split. This PR only edits two pre-existing files in it. Not actionable here without a repo-wide guide-layout change.
